HeterogeneousCore/CUDAUtilities/interface/HistoContainer.h

```diff
@@ -0,0 +1,161 @@
+
+
+#include<cassert>
+#include<cstdint>
+#include<algorithm>
+#ifndef __NVCC__
+#include<atomic>
+#endif
+#ifdef __NVCC__
+#include <cuda_runtime.h>
+#endif
+
+#include "HeterogeneousCore/CUDAUtilities/interface/cudastdAlgorithm.h"
+
+#ifdef __NVCC__
+
+  template<typename Histo>
+  __host__
+  void zero(Histo * h, uint32_t nh, int nthreads, cudaStream_t stream) {
+    auto nblocks = (nh*Histo::nbins()+nthreads-1)/nthreads;
+    zeroMany<<<nblocks,nthreads, 0, stream>>>(h,nh);
+  }
+
+  template<typename Histo, typename T>
+  __host__
+  void fillOneFromVector(Histo * h, T const * v, uint32_t size, int nthreads, cudaStream_t stream) {
+    zero(h,1, nthreads, stream);
+    auto nblocks = (size+nthreads-1)/nthreads;
+    fillFromVector<<<nblocks,nthreads, 0, stream>>>(h,v,size);
+  }
+
+  template<typename Histo, typename T>
+  __host__
+  void fillManyFromVector(Histo * h, uint32_t nh, T const * v, uint32_t * offsets, uint32_t totSize, int nthreads, cudaStream_t stream) {
+    zero(h,nh, nthreads, stream);
+    auto nblocks = (totSize+nthreads-1)/nthreads;
+    fillFromVector<<<nblocks,nthreads, 0, stream>>>(h,nh,v,offsets);
+  }
+
+  template<typename Histo>
+  __global__
+  void zeroMany(Histo * h, uint32_t nh) {
+    auto i = blockIdx.x*blockDim.x + threadIdx.x;
+    auto ih = i/Histo::nbins();
+    auto k = i - ih*Histo::nbins();
+    if (ih<nh) {
+      h[ih].nspills=0;
+      if(k<Histo::nbins()) h[ih].n[k]=0;
+    }
+  }
+
+  template<typename Histo, typename T>
+  __global__
+  void fillFromVector(Histo * h,  uint32_t nh, T const * v, uint32_t * offsets) {
+     auto i = blockIdx.x*blockDim.x + threadIdx.x;
+     if(i>=offsets[nh]) return;
+     auto off = cuda_std::upper_bound(offsets,offsets+nh+1,i);
+     assert((*off)>0);
+     int32_t ih = off-offsets-1;
+     assert(ih>=0);
+     assert(ih<nh); 
+     h[ih].fill(v,i);
+  }
+
+
+  template<typename Histo, typename T>
+  __global__
+  void fillFromVector(Histo * h, T const * v, uint32_t size) {
+     auto i = blockIdx.x*blockDim.x + threadIdx.x;
+     if(i<size) h->fill(v,i);
+  }
+
+#endif
+
+template<typename T, uint32_t N, uint32_t M>
+class HistoContainer {
+public:
+  static constexpr uint32_t sizeT() { return sizeof(T)*8; }
+  static constexpr uint32_t nbins() { return 1<<N; }
+  static constexpr uint32_t shift() { return sizeT() -N; }
+  static constexpr uint32_t mask() { return  nbins()-1; }
+  static constexpr uint32_t binSize() { return 1<<M; }
+  static constexpr uint32_t spillSize() { return 4*binSize(); }
+
+  static constexpr uint32_t bin(T t) {
+    return (t>>shift())&mask();
+  }
+
+
+#ifdef __NVCC__
+
+  __device__
+  void fill(T const * t, uint32_t j) {
+    auto b = bin(t[j]);
+    auto w = atomicAdd(&n[b],1); // atomic
+    if (w<binSize()) {
+      bins[b*binSize()+w] = j;
+    } else {
+      auto w = atomicAdd(&nspills,1); // atomic
+      if (w<spillSize()) spillBin[w] = j;
+    }
+  }     
+
+#else
+  
+  HistoContainer() { zero();}
```

Comment:
on the GPU, would you rather have the default constructor, or no public constructor ?

Reply:
Do not remember which error was giving: I think it needs the default constructor.

Reply:
I think the present code will not do what you expect: when building CUDA code on the host, `__NVCC__` is defined, so the compiler will not see the explicit constructor and generate a default one, which will not call `zero()`.

I would suggest to `#ifdef` the constructor based on __CUDA_ARCH__, so that
  - non CUDA code will see the user-defined constructor;
  - CUDA code on the host will see the user-defined constructor;
  - CUDA code on the device will *not* see the user-defined constructor, and fall back to the default one.

Reply:
Sigh... while it should work as far as CUDA is concerned, the cuda-api-wrappers library tries to be helpful and complains about a non-trivial constructor in device code - even though the constructor is non-trivial only on the host.

Reply:
so adding __host__ and __device__ in front of the constructor
(actually just adding __device__ HistoContainer() = default; )
in any case I think that in .cu i do not zero() o host I zero on device
zero(Histo * h, uint32_t nh, int nthreads, cudaStream_t stream);

HistoContainer() { zero();} is needed only in the CPU implementation